library/alloc/src/rc.rs

```diff
@@ -386,7 +386,35 @@ impl<T> Rc<T> {
     /// returns, calling [`upgrade`] on the weak reference inside `data_fn` will
     /// fail and result in a `None` value.
     ///
+    /// # Why using this method?
+    ///
+    /// It allows to construct a `T` which holds a `Weak<T>` pointer to itself.
+    /// Without this method, you would need to use [`Weak::dangling`] and then
+    /// mutate the value by using [`Rc::get_mut`] to fix up the self references.
+    ///
+    /// # Why is a closure needed?
+    ///
+    /// The place needs to be "finished" once the value is created. In theory,
+    /// you could "recover" any zombie weak, but for the most case, Rust wants
+    /// to pretend that a "zombie" [`Weak`] and a dangling [`Weak`] are
+    /// equivalent.
+    ///
+    /// # What is the closure argument about?
+    ///
+    /// The point of `new_cyclic` is that the place is allocated first, with
+    /// `strong=0` and `weak=1`. The place has no data yet (you're constructing
+    /// the value to place there), but is in the "zombie" state where it has
+    /// weak references but no data.
+    ///
+    /// # What should the closure return?
+    ///
+    /// The closure returns the object to place in the reference counted place.
+    /// After you've constructed the object, the reference count is "activated"
+    /// and the object is accessable by (clones of) the weak pointer given to the
+    /// constructor closure.
```

Comment:
IMO most of this fits better as actual `//` comments in the source. If users are interested in this kind of detail, they'll likely open the source anyway.

I tend to prefer to explain these things with documentation inside the doc example. For example:
```rust
impl Gadget {
    /// Construct a reference counted Gadget.
    fn new() -> Rc<Self> {
        // `me` is an unitialized Weak at the start of the closure
        Rc::new_cyclic(|me| {
            // create the actual struct here
            Gadget { me: me.clone() }
            
            // before returning, `new_cyclic` fixes up the pointers
            // and reference counts inside gadget
        })
    }
}
```

Reply:
This a very good suggestion. I'll apply it then.

Reply:
Actually I'll keep both. Some extra explanations doesn't hurt. :)